### PR TITLE
XWIKI-13470: Allow multiselect for livetable list filters

### PR DESCRIPTION
**Pull request by Pbast**

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 1]
```diff
@@ -764,28 +775,35 @@
...
 #end
 
 
-#macro (livetable_getFilterQuery $column $matchType $isList)
+#macro (livetable_getFilterQuery $column $matchType $isList $multilistsize)
   #if ($matchType == 'partial' || $matchType == 'prefix')
     upper($column) like upper(?)
   #elseif ($isList)
     ? in elements($column)
+  #elseif (($multilistsize > 1) && ($matchType == 'exact'))
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: ``&&`` is not encoded. Same below. You should avoid modifying the page XML directly.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 2]
```diff
@@ -764,28 +775,35 @@
...
+    $column in (? , ?)
+    $column in ( $stringtool.repeat('?', ', ', $multilistsize) )
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: There seems to be some duplication here.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 3]
```diff
@@ -764,28 +775,35 @@
...
 
 
-#macro (livetable_addFilterParam $filterValue $matchType $params)
+#macro (livetable_addFilterParam $filterValue $matchType $params $multilistsize)
   #if ($matchType == 'partial')
     #set ($discard = $params.add("%$!filterValue%"))
   #elseif ($matchType == 'prefix')
     #set ($discard = $params.add("$!filterValue%"))
+  #elseif (($multilistsize > 1) && ($matchType == 'exact'))
+    #foreach ($filterValue in $filterValues)
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Where does ``$filterValues`` come from? The macro parameter is ``$filterValue``. Also, overwriting macro parameters is considered a bad practice in general. It may have unexpected side effects.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 1 above)
Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Fixed.
"avoid modifying the page XML directly" what do you mean? I "git checkout" source to modify sources files (with vi).
Am I wrong?

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 2 above)
Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: encoding fixed too. thxs

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 3 above)
Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Yes $filterValues is an array ( different from $filterValue containing a string)
$filterValues come from
Line 632
```
        #set ($filterValues = $request.getParameterValues($colname))
	#set ($multilistsize = $filterValues.size())
```
Then, do you propose to replace $filterValue by $filterValues (an array variable with one or multiple value) everywhere in LiveTableResultsMacros?

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 1 above)
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: > Am I wrong?

Yes. Wiki pages should be modified on a running XWiki instance, then exported, then copied to the source tree, and then formatted with ``mvn xar:format``. If you modify wiki pages directly in the source tree then:
* you always forget about XML encoding
* you don't see the effect of your changes on a running XWiki instance.

Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: yes but the problem is "mvn xar:format" doesn't work and I didn't understand why (I don't know enough Maven).

I launch
```
cd /home/git-wiki/xwiki-platform/xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui
mvn xar:format
```
and I obtain:
```
INFO] --- xwiki-commons-tool-xar-plugin:8.2-SNAPSHOT:format (default-cli) @ xwiki-platform-livetable-ui ---
juin 10, 2016 2:07:26 PM org.sonatype.guice.bean.reflect.Logs$JULSink warn
AVERTISSEMENT: Error injecting: org.xwiki.tool.xar.FormatMojo
java.lang.TypeNotPresentException: Type org.xwiki.tool.xar.FormatMojo not present
....
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 7.622s
[INFO] Finished at: Fri Jun 10 14:07:26 CEST 2016
[INFO] Final Memory: 21M/232M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.xwiki.commons:xwiki-commons-tool-xar-plugin:8.2-SNAPSHOT:format (default-cli) on project xwiki-platform-livetable-ui: Execution default-cli of goal org.xwiki.commons:xwiki-commons-tool-xar-plugin:8.2-SNAPSHOT:format failed: Unable to load the mojo 'format' in the plugin 'org.xwiki.commons:xwiki-commons-tool-xar-plugin:8.2-SNAPSHOT' due to an API incompatibility: org.codehaus.plexus.component.repository.exception.ComponentLookupException: org/xwiki/tool/xar/FormatMojo : Unsupported major.minor version 52.0
[ERROR] -----------------------------------------------------
[ERROR] realm =    plugin>org.xwiki.commons:xwiki-commons-tool-xar-plugin:8.2-SNAPSHOT
[ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
[ERROR] urls[0] = file:/root/.m2/repository/org/xwiki/commons/xwiki-commons-tool-xar-plugin/8.2-SNAPSHOT/xwiki-commons-tool-xar-plugin-8.2-SNAPSHOT.jar
[ERROR] urls[1] = file:/root/.m2/repository/dom4j/dom4j/1.6.1/dom4j-1.6.1.jar

```

Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: > Unsupported major.minor version 52.0

Looks like you are not using the right Java version. ``xwiki-commons-tool-xar-plugin:8.2-SNAPSHOT`` has been build on our CI agents using Java 8 and you are trying to use it with Java 7 or 6.

Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Ha right. I use an old debian (wheezy with Java7)...
I downloaded/installed a fresh jdk and ... 

```
[INFO] Updating license headers...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
```

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 4]
```diff
@@ -208,12 +208,18 @@
       #set($discard = $allMatchingParams.addAll($filterParams))
     #end
   #end
+  #if ($request.debug) ##Mode Debug
```
Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: There is already a debug output at [line 425](https://github.com/Pbast/xwiki-platform/blob/2c1f0b5d85d25bdce3ffcbba111cd7f96f79516f/xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml#L425), why don't you extend that one?

File: xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/macros.vm [hunk 5]
```diff
@@ -277,7 +277,13 @@
                 #set($class = $xwiki.getDocument($propClassName).getxWikiClass())
                 #set($list = '')
                 #set($list = $class.get($colname).getListValues())
-                <select id="xwiki-livetable-${htmlLiveTableId}-filter-${velocityCount}" name='${colname}'><option value=''>$services.localization.render('platform.livetable.selectAll')</option>
+                <select id="xwiki-livetable-${htmlLiveTableId}-filter-${velocityCount}" name='${colname}'
+                #if("$!colprop.type" == 'multilist')
+                  class='xwiki-livetable-multilist' multiple='multiple'>
+                #else
+                  ><option value=''>$services.localization.render('platform.livetable.selectAll')</option>
+                  <option disabled>────</option>
```
Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/macros.vm]: Inconsistent use of valueless attributes, 3 lines above you used `disabled='disabled'`.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 6]
```diff
@@ -623,7 +629,12 @@
           #set ($discard = $whereParams.add("%$filterValue%"))
         #end
       #else
-        #livetable_filterProperty($colName $filterValue)
+        #set ($filterValues = $request.getParameterValues($colname))
+	#set ($multilistsize = $filterValues.size())
```
Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Don't use tabs for indentation.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 4 above)
Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Because I didn't find this one (I add this in documentation: http://extensions.xwiki.org/xwiki/bin/view/Extension/Livetable+Macro#HHowtodebuganissuewiththeLivetabledisplay)

File: xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/macros.vm (hunk 5 above)
Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/macros.vm]: fixed

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 6 above)
Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: fixed too. thxs

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 7]
```diff
@@ -764,28 +767,35 @@
...
 
-#macro (livetable_getFilterQuery $column $matchType $isList)
+#macro (livetable_getFilterQuery $column $matchType $isList $multilistsize)
   #if ($matchType == 'partial' || $matchType == 'prefix')
     upper($column) like upper(?)
   #elseif ($isList)
     ? in elements($column)
+  #elseif (($multilistsize &gt; 1) &amp;&amp; ($matchType == 'exact'))
+    $column in (? , ?)
+    $column in ( $stringtool.repeat('?', ', ', $multilistsize) )
```
Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: These two lines seem wrong to me. Can you double check?

Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: (I'm home now without my files, git and dev tools).
Maybe I uncommented by mistake Line782 ?
I will add this file in my xarf  file, import it and test it monday

Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: fixed (you are right).
This time, I tested, export xar file and launch "mvn xar:format" but xml file have 1.2 version instead 1.1.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 3 above)
Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Fixed (delete this bloc)
```
        #if ($multilistsize > 1)
          #set($filterValue=$filterValues)
        #end
```

**Comment by Pbast** [comment]
Do you have any comments?
Thxs

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 8]
```diff
@@ -30,8 +30,8 @@
   <parent>XWiki.WebHome</parent>
   <author>xwiki:XWiki.Admin</author>
   <contentAuthor>xwiki:XWiki.Admin</contentAuthor>
-  <date>1465822369000</date>
-  <contentUpdateDate>1465822369000</contentUpdateDate>
+  <date>1465895028000</date>
```
Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: This is noise added to every commit, so it shouldn't be committed. Do you know about `git add --interactive`?

**Comment by sdumitriu** [comment]
:+1:

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 8 above)
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: I don't agree Sergiu. I also commit this usually because it makes sense to me that when we make a modification the date is that of the modification and not keep the old date. Now that said we had discussions about this in the past but I don't think that the XAR plugin is resetting any date ATM (would need to check).

Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: While I agree that the date of a document should be updated from time to time to reflect when it was actually updated in the sources, I find that including changes like this in every commit simply adds noise distracting from the review. At the very least, it should be done in a separate commit.

Comment (sdumitriu) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: But this is just a philosophical discussion, @Pbast you can ignore this for the moment.

Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: FTR, I agree with Sergiu. I'm a bit surprised since I was pretty sure this was the best practice. At least that's what I always do (and what I've seen the others do).

Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: FTR I've found http://jira.xwiki.org/browse/XWIKI-11764 which is still open.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 9]
```diff
@@ -764,28 +763,34 @@
...
+  #elseif (($multilistsize &gt; 1) &amp;&amp; ($matchType == 'exact'))
+    $column in ( $stringtool.repeat('?', ', ', $multilistsize) )
   #else
     $column = ?
   #end
 #end
 
 
-#macro (livetable_addFilterParam $filterValue $matchType $params)
+#macro (livetable_addFilterParam $filterValue $matchType $params $multilistsize)
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: I wouldn't modify this macro. I would do the ``#foreach`` outside. This way it will work with any ``$matchType`` value.

    #foreach ($filterValue in $filterValues)
      #livetable_addFilterParam($filterValue $matchType $whereParams)
    #end

Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: I wonder if your loop won't have impact with:
#macro (livetable_filterDBStringListProperty)
Anyway, there are same loop in this macro,  I will give a try

**Comment by mflorea** [comment]
@Pbast note that there is a ``macros.vm`` file in ``xwiki-platform-web`` also (Flamingo skin is overwriting it). You need to update it also.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 9 above)
Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Nope, your code failed if I use your loop :-(
Maybe I made a mistake but computed query must have OR condition field IN (?,?,xx) and not AND

If you prefer, I can remove  $multilistsize parameter from
`#livetable_filterProperty($colName $filterValue $multilistsize)`
because

> NOTE: This macro uses variables defined in livetable_filterProperty . It was not meant to be used alone.

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml [hunk 10]
```diff
@@ -764,28 +763,34 @@
...
-  #set ($filterQuery = "#livetable_getFilterQuery(""${safe_tableAlias}.value"" $matchType)")
+  #set ($filterQuery = "#livetable_getFilterQuery(""${safe_tableAlias}.value"" $matchType false $multilistsize)")
   #set ($whereSql = "${whereSql} and $filterQuery.trim()")
-  #livetable_addFilterParam($filterValue $matchType $whereParams)
+  #livetable_addFilterParam($filterValue $matchType $whereParams $multilistsize)
 #end
 
 
-#macro (livetable_getFilterQuery $column $matchType $isList)
+#macro (livetable_getFilterQuery $column $matchType $isList $multilistsize)
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: I would rename ``$multilistsize`` to ``$valueCount`` (or something similar, camel case anyway) in order to make it more generic. For instance, when ``$valueCount`` is ``1``

    $column in (?)

is equivalent to

    $column = ?

And in the future we may want to add support for matching against multiple values even for the *partial* and *prefix* matching type

    ($stringtool.repeat("upper($column) like upper(?)", ' or ', $valueCount))

**Comment by Pbast** [comment]
@mflorea yes thxs I know, I'm waiting to fix definitively macros.vm from flamingo skin to merge modifications in the other macros.vm :-)

File: xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml (hunk 9 above)
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: From what I can see ``livetable_addFilterParam`` only adds parameters to the ``$params`` so no AND/OR is needed. Calling ``livetable_addFilterParam`` in a loop should add multiple parameters to ``$params``. When I said

> do the ``#foreach`` outside

I meant right when ``livetable_addFilterParam`` is called. Now, in order to not do the ``#foreach`` loop every time ``livetable_addFilterParam`` is called, we could define ``livetable_addFilterParams`` (plural).

Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: Ok I get it.
Previously I modified wrong macro :-/

Comment (Pbast) [inline xwiki-platform-core/xwiki-platform-livetable/xwiki-platform-livetable-ui/src/main/resources/XWiki/LiveTableResultsMacros.xml]: fixed :-)

**Comment by mflorea** [comment]
@Pbast thank you for this PR and for your patience :)

**Comment by Pbast** [comment]
ha ha It was a test, isn't it?

More seriously, thxs for your comments to clean my messy code.
